File: 2024/March/3-3 19 Remove Nth Node From End of List/Approach1.java
public class Approach1 {
    public static void main(String[] args) {

    }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode node = head;
        int length = 0;

        while(node != null) {
            length++;
            node = node.next;
        }

        ListNode previous = null;
        node = head;
        for(int i = length - n; i > 0; i--) {
            previous = node;
            node = node.next;
        }

        if(previous == null) {
            return node.next;
        }
        else {
            previous.next = node.next;
        }

        return head;
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

File: 2024/March/3-3 19 Remove Nth Node From End of List/Approach1.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        node = head
        length = 0
        
        while node:
            length += 1
            node = node.next
        
        previous = None
        node = head
        for i in range(length - n):
            previous = node
            node = node.next

        if not previous:
            return node.next
        else:
            previous.next = node.next
    
        return head